0x06c Restructure basic information gathering (#1816)
OWASP/owasp-mstg: Document/0x06c-Reverse-Engineering-and-Tampering.md

$ otool -L <binary>
```

#### Retrieving Strings

> Strings are always a good starting point while analyzing a binary, as it provides a context to the associated code. For instance, an error log string "Cryptogram generation failed." gives us a hint that the code might be responsible for generation of a cryptogram.

Reviewer: ```suggestion
Strings are always a good starting point while analyzing a binary, as they provide context to the associated code. For instance, an error log string such as "Cryptogram generation failed" gives us a hint that the adjoining code might be responsible for the generation of a cryptogram.
```

> Strings are always a good starting point while analyzing a binary, as it provides a context to the associated code. For instance, an error log string "Cryptogram generation failed." gives us a hint that the adjoining code might be responsible for generation of a cryptogram.

Author: Thanks for that. This is helpful for me 👍

> Strings can be extracted from an iOS binary using a GUI tools like Ghidra or Cutter, using a CLI-based tool such as the *strings* Unix utility (`strings <path_to_binary>`) or radare2's rabin2 (`rabin2 -zz <path_to_binary>`). When using the CLI-based tools, you use them in conjunction with other tools like *grep* or *sort* to further filter and analyze the results.

Reviewer: This paragraph comes from the Android part of the guide, if it's anyway the same let's leave it as it was.

```suggestion
In order to extract strings from an iOS binary, you can use GUI tools such as Ghidra or Cutter or rely on CLI-based tools such as the _strings_ Unix utility (`strings <path_to_binary>`) or radare2's rabin2 (`rabin2 -zz <path_to_binary>`). When using the CLI-based ones you can take advantage of other tools such as grep (e.g. in conjunction with regular expressions) to further filter and analyze the results.
```

> Strings can be extracted from an iOS binary using a GUI tools like Ghidra or Cutter, or using a CLI-based tool such as *strings* Unix utility (`strings <path_to_binary>`) or radare2's rabin2 (`rabin2 -zz <path_to_binary>`). When using the CLI-based tools, one can use it in conjunction with other tools like *grep* or *sort* to further filter and analyze the results.

Reviewer 2: ```suggestion
Strings can be extracted from an iOS binary using a GUI tools like Ghidra or Cutter, using a CLI-based tool such as the *strings* Unix utility (`strings <path_to_binary>`) or radare2's rabin2 (`rabin2 -zz <path_to_binary>`). When using the CLI-based tools, you use them in conjunction with other tools like *grep* or *sort* to further filter and analyze the results.
```

#### Cross References

> Ghidra can be used for analyzing the iOS binaries and obtain cross references by right clicking the desired function and selecting **Show References to**.

Reviewer 2: ```suggestion
Ghidra can be used for analyzing the iOS binaries and obtaining cross references by right clicking the desired function and selecting **Show References to**.
```
Radare2 probably also has this, no?

Author: Could be. I haven't used R2, and thus I avoid mentioning it.

#### API Usage

> The iOS platform provides many in-built libraries for frequently used functionalities in applications, for example cryptography, Bluetooth, NFC, network or location libraries. Determining the presence of these libraries in an application can give us valuable information about its underlying working.

Reviewer 2: ```suggestion
The iOS platform provides many built-in libraries for frequently used functionalities in applications, for example cryptography, Bluetooth, NFC, network and location libraries. Determining the presence of these libraries in an application can give us valuable information about its underlying working.
```

> For instance, if an application is importing `CC_SHA256` function, it indicates the application will be performing some kind of hashing operation using SHA256 algorithm. Further information on how to analyze iOS's cryptographic APIs is discussed in the section "[iOS Cryptographic APIs](0x06e-testing-cryptography "iOS Cryptographic APIs")".

Reviewer 2: ```suggestion
For instance, if an application is importing the `CC_SHA256` function, it indicates that the application will be performing some kind of hashing operation using the SHA256 algorithm. Further information on how to analyze iOS's cryptographic APIs is discussed in the section "[iOS Cryptographic APIs](0x06e-testing-cryptography "iOS Cryptographic APIs")".
```

> Similarly, the above approach can be used to determine where and how an application is using Bluetooth. For instance, an application performing communication using Bluetooth channel must use functions from Core Bluetooth framework like `CBCentralManager` or `connect`. Using the [iOS Bluetooth documentation](https://developer.apple.com/documentation/corebluetooth "iOS Bluetooth documenation") one can determine the critical functions and start analysis around such function imports.

Reviewer 2: ```suggestion
Similarly, the above approach can be used to determine where and how an application is using Bluetooth. For instance, an application performing communication using the Bluetooth channel must use functions from the Core Bluetooth framework such as `CBCentralManager` or `connect`. Using the [iOS Bluetooth documentation](https://developer.apple.com/documentation/corebluetooth "iOS Bluetooth documenation") you can determine the critical functions and start analysis around those function imports.
```

#### Network Communication

Most of the apps you might encounter connect to remote endpoints. Even before you perform any dynamic analysis (e.g. traffic capture and analysis), you can obtain some initial inputs or entry points by enumerating the domains to which the application is supposed to communicate to.

> Typically these domains will be present as strings within the binary of the application. One can extract domains by retrieving strings (as discussed above) or checking the strings using tools like Ghidra. The latter option has a clear advantage: it can provide you with context, as you'll be able to see in which context each domain is being used by checking the cross-references.

Reviewer 2: This one's a bit weird. The section above explains that you can get strings with both (strings command and gui stuff), and the cross-reference section is on the same level as 'retrieving strings', so not really related to it. Maybe we should remove the cross-reference titel (and thus link it to the strings part) and then this advantage can also be put there.

Author: Currently its there to keep iOS and Android structure same, something pointed out to me in past. Given there is only a single line of content under "cross-references" heading, I dont mind removing it. @cpholguera what is your opinion on this one?

Reviewer: The Cross References section should remain since it's an important part and it's also on the Android chapter. In this chapter it should highlight what's different/special in comparison to Android (in contrast). How's following xrefs in Swift vs objc vs native? maybe a bit rephrasing would help. The point here is to tell how you find an interesting string and use "data xrefs" to get where you want, see in which context is used, from there you'd keep using "code xrefs" to get to other related functions. I guess that's what you wanted to transmit here and in the xrefs section. Was it?

Reviewer: Can you use Ghidra to plot a xref data graph? that would be nice to show this point, showing where a string is being used, in this case an URL string. I know you can do it with radare2, but I'm not sure with Ghidra.

Author: I kept this sub-section aligned with android. And since the section is about "basic information gathering" I kept it simple. Ghidra is used in other places in the chapter after this section. 

The example of using xref to determine where the strings is used, is demonstrated in "Reviewing Disassembled Objective-C and Swift Code" section.  

I can expand this section more, but i feel the information might be repetitive in nature.

From here on you can use this information to derive more insights which might be of use later during your analysis, e.g. you could match the domains to the pinned certificates or perform further reconnaissance on domain names to know more about the target environment.

> The implementation and verification of secure connections can be an intricate process and there are numerous aspects to consider. For instance, many applications use other protocols apart from HTTP such as XMPP or plain TCP packets, or perform certificate pinning in an attempt to deter MITM attacks but unfortunately having severe logical bugs in its implementation.

Reviewer 2: ```suggestion
The implementation and verification of secure connections can be an intricate process and there are numerous aspects to consider. For instance, many applications use other protocols apart from HTTP such as XMPP or plain TCP packets, or perform certificate pinning in an attempt to deter MITM attacks.
```
I would remove the last part, since it's not really necessary here and it makes the sentence quite long and convoluted.

> Remember that in most of the cases, just using static analysis will not be enough and might even turn to be extremely inefficient when compared to the dynamic alternatives which will get much more reliable results (e.g. using an interceptor proxy). In this section we've just slightly touched the surface, please refer to the section "[Basic Network Monitoring/Sniffing](0x06b-Basic-Security-Testing.md#basic-network-monitoringsniffing "Basic Network Monitoring/Sniffing")" in the "iOS Basic Security Testing" chapter and also check the test cases in the chapter "[iOS Network APIs](0x06g-testing-network-communication "iOS Network APIs")".

Reviewer 2: ```suggestion
Remember that in most cases, only using static analysis will not be enough and might even turn to be extremely inefficient when compared to the dynamic alternatives which will get much more reliable results (e.g. using an interceptor proxy). In this section we've just slightly touched the surface, so please refer to the section "[Basic Network Monitoring/Sniffing](0x06b-Basic-Security-Testing.md#basic-network-monitoringsniffing "Basic Network Monitoring/Sniffing")" in the "iOS Basic Security Testing" chapter and check out the test cases in the chapter "[iOS Network APIs](0x06g-testing-network-communication "iOS Network APIs")" for further information.
```
